tests/testflows/map_type/requirements/requirements.md

```diff
@@ -0,0 +1,523 @@
+All the updates are tracked using the [Revision History].
+
+## Introduction
+
+This software requirements specification covers requirements for `Map(key, value)` data type in [ClickHouse].
+
+## Requirements
+
+### General
+
+#### RQ.SRS-018.ClickHouse.Map.DataType
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type that stores `key:value` pairs.
+
+### Performance
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Performance.Vs.ArrayOfTuples
+version:1.0
+
+[ClickHouse] SHALL provide comparable performance for `Map(key, value)` data type as
+compared to `Array(Tuple(K,V))` data type.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Performance.Vs.TupleOfArrays
+version:1.0
+
+[ClickHouse] SHALL provide comparable performance for `Map(key, value)` data type as
+compared to `Tuple(Array(String), Array(String))` data type where the first
+array defines an array of keys and the second array defines an array of values.
+
+### Key Types
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Key.String
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type where key is of a [String] type.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Key.Integer
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type where key is of an [Integer] type.
+
+### Value Types
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.String
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type where value is of a [String] type.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Integer
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type where value is of a [Integer] type.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Array
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type where value is of a [Array] type.
+
+### Invalid Types
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Invalid.Nullable
+version: 1.0
+
+[ClickHouse] SHALL not support creating table columns that have `Nullable(Map(key, value))` data type.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Invalid.NothingNothing
+version: 1.0
+
+[ClickHouse] SHALL not support creating table columns that have `Map(Nothing, Nothing))` data type.
+
+### Duplicated Keys
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.DuplicatedKeys
+version: 1.0
+
+[ClickHouse] SHALL support `Map(key, value)` data type with duplicated keys.
+
+### Array of Maps
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.ArrayOfMaps
+version: 1.0
+
+[ClickHouse] SHALL support `Array(Map(key, value))` data type.
+
+### Nested With Maps
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.NestedWithMaps
+version: 1.0
+
+[ClickHouse] SHALL support defining `Map(key, value)` data type inside the [Nested] data type.
+
+### Value Retrieval
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Retrieval
+version: 1.0
+
+[ClickHouse] SHALL support getting the value from a `Map(key, value)` data type using `map[key]` syntax.
+If `key` has duplicates then the first `key:value` pair SHALL be returned. 
```

Comment:
Why first?
I prefer to left it unspecified.

Reply:
Same as above. This is the current behavior and if it changes in the future we can update this requirement. Having basic corner cases being unspecified is not a good practice.

Reply:
Updated the requirement to use MAY.